### DISCUSSION TITLE
How does ripgrep exclude binary files?

### DISCUSSION THREAD

**Discussion by d3lm** [BurntSushi/ripgrep#2050] [Q&A] [answered]
I was wondering how ripgrep excludes binary files. Does it use something like `from_utf8` and if it fails it won't further process the file? I'd imagine that's not the case because it's very optimized and performs search on raw bytes.

The reason I am asking this is because ripgrep is used by VSCode and it's not returning file matches within binary files. I was wondering if that's something ripgrep does or if VSCode maybe passes some exclude pattern.

**Comment by d3lm** [comment, marked as answer]
Oh I think I actually just found the answer [here](https://github.com/BurntSushi/ripgrep/blob/master/GUIDE.md#binary-data). It looks for a `NUL` byte.

**Reply by BurntSushi** [reply]
Yup that is indeed correct! I wrote that section because binary data detection is quite complicated in how it mingles with the various filtering strategies and search strategies.

And note that the `-a/--text` flag can be passed which will disable all binary detection whatsoever.

**Reply by d3lm** [reply]
Yep, thanks a lot! For my smaller ripgrep clone I am using `from_utf8` actually even though that's also not 100% correct, but I figured it might be a bit better than just checking for a `NUL` byte.
